src/main/java/org/hydev/mcpm/client/database/DatabaseInteractor.java

```diff
@@ -88,6 +101,35 @@ public ListPackagesResult list(ListPackagesInput input) {
         );
     }
 
+    /**
+     * Searches for plugins based on the provided name, keyword, or command.
+     * The input contains the type of search.
+     *
+     * @param input Record of inputs as provided in SearchPackagesInput. See it for more info.
+     * @return Packages result. See the SearchPackagesResult record for more info.
+     */
+    public SearchPackagesResult search(SearchPackagesInput input) {
+        var database = fetcher.fetchDatabase(!input.noCache(), listener);
+
+        if (database == null) {
+            return SearchPackagesResult.by(SearchPackagesResult.State.FAILED_TO_FETCH_DATABASE);
+        }
+
+        var searchStr = input.searchStr();
+        if (searchStr.isEmpty())
+            return SearchPackagesResult.by(SearchPackagesResult.State.INVALID_INPUT);
+
+        var plugins = database.plugins();
+
+        // Make map if null
+        SearcherFactory factory = new SearcherFactory();
+        if (this.searchMaps.get(input.type()) == null)
+            this.searchMaps.put(input.type(), factory.createSearcher(input).constructSearchMaps(plugins));
```

Comment:
Here, you create a factory even if your index for that use case is populated.
But you avoid creating a factory by createSearcher a static method.

Although it seems to me that the searcher objects themselves should be able to have the search method as static, static methods defined in an interface cannot be overriden which wouldn't work.

Reply:
createSearcher being static makes sense, and it's what i had. But a brief glance online had some ppl doing the factory design pattern without static. ig ill go with static since it makes more sense.